uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,794 @@
+  T ReadAt(const std::function<T(size_t)>& aClosure) {
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += sizeof(T);
+    AssertInBounds(newOffset);
+    T result = aClosure(mOffset.value());
+    mOffset = newOffset;
+    return result;
+  }
+
+  const RustBuffer& mBuffer;
+  CheckedInt<size_t> mOffset;
+};
+
+class MOZ_STACK_CLASS Writer final {
+ public:
+  explicit Writer(size_t aCapacity) : mBuffer(aCapacity) {}
+
+  void WriteUInt8(const uint8_t& aValue) {
+    WriteAt<uint8_t>(aValue, [this](size_t aOffset, const uint8_t& aValue) {
+      mBuffer[aOffset] = aValue;
+    });
+  }
+
+  void WriteInt8(const int8_t& aValue) {
+    WriteUInt8(BitwiseCast<uint8_t>(aValue));
+  }
+
+  // This code uses `memcpy` and other eye-twitchy patterns because it
+  // originally wrote values directly into a `RustBuffer`, instead of
+  // an intermediate `nsTArray`. Once #251 is fixed, we can return to
+  // doing that, and remove `ToRustBuffer`.
+
+  void WriteUInt16(const uint16_t& aValue) {
+    WriteAt<uint16_t>(aValue, [this](size_t aOffset, const uint16_t& aValue) {
+      uint16_t value = PR_htons(aValue);
+      memcpy(&mBuffer.Elements()[aOffset], &value, sizeof(uint16_t));
+    });
+  }
+
+  void WriteInt16(const int16_t& aValue) {
+    WriteUInt16(BitwiseCast<uint16_t>(aValue));
+  }
+
+  void WriteUInt32(const uint32_t& aValue) {
+    WriteAt<uint32_t>(aValue, [this](size_t aOffset, const uint32_t& aValue) {
+      uint32_t value = PR_htonl(aValue);
+      memcpy(&mBuffer.Elements()[aOffset], &value, sizeof(uint32_t));
+    });
+  }
+
+  void WriteInt32(const int32_t& aValue) {
+    WriteUInt32(BitwiseCast<uint32_t>(aValue));
+  }
+
+  void WriteUInt64(const uint64_t& aValue) {
+    WriteAt<uint64_t>(aValue, [this](size_t aOffset, const uint64_t& aValue) {
+      uint64_t value = PR_htonll(aValue);
+      memcpy(&mBuffer.Elements()[aOffset], &value, sizeof(uint64_t));
+    });
+  }
+
+  void WriteInt64(const int64_t& aValue) {
+    WriteUInt64(BitwiseCast<uint64_t>(aValue));
+  }
+
+  void WriteFloat(const float& aValue) {
+    WriteUInt32(BitwiseCast<uint32_t>(aValue));
+  }
+
+  void WriteDouble(const double& aValue) {
+    WriteUInt64(BitwiseCast<uint64_t>(aValue));
+  }
+
+  /// Writes a length-prefixed UTF-8 encoded string at the current offset. The
+  /// closure takes a `Span` pointing to the byte buffer, which it should fill
+  /// with bytes and return the actual number of bytes written.
+  ///
+  /// This function is (more than a little) convoluted. It's written this way
+  /// because we want to support UTF-8 and UTF-16 strings. The "size hint" is
+  /// the maximum number of bytes that the closure can write. For UTF-8 strings,
+  /// this is just the length. For UTF-16 strings, which must be converted to
+  /// UTF-8, this can be up to three times the length. Once the closure tells us
+  /// how many bytes it's actually written, we can write the length prefix, and
+  /// advance the current offset.
```

Comment:
Why does `ReadRawString` only support UTF-8 when this supports UTF-8 and UTF-16?

Reply:
It supports reading into a UTF-8 and a UTF-16 string, too—but the bytes in the `RustBuffer` are always UTF-8 encoded, to match Rust strings. But I realized the signatures for `ReadRawString` and `WriteRawString` were super confusing, especially with the lambda doing strange things, so I flattened them into `ReadCString` and `ReadString` methods. It's a little more code, but a lot more readable, so seems like worth the trade-off.